internal/config/network/config.go

```diff
@@ -25,32 +27,100 @@ type TLS struct {
 	Key []byte
 	//Certificate for the TLS connection (CertPath will this overwrite, if specified)
 	Cert []byte
+
+	// cachedCert and cachedCertModTime are used to cache the parsed certificate
+	cachedCert        *tls.Certificate
+	cachedCertModTime time.Time
+	cachedKeyModTime  time.Time
+	mtx               *sync.RWMutex
 }
 
-func (t *TLS) Config() (_ *tls.Config, err error) {
-	if !t.Enabled {
-		return nil, nil
+func (t *TLS) getCert() ([]byte, error) {
+	if t.CertPath != "" {
+		return os.ReadFile(t.CertPath)
 	}
+	return t.Cert, nil
+}
+
+func (t *TLS) getKey() ([]byte, error) {
 	if t.KeyPath != "" {
-		t.Key, err = os.ReadFile(t.KeyPath)
-		if err != nil {
-			return nil, err
-		}
+		return os.ReadFile(t.KeyPath)
 	}
-	if t.CertPath != "" {
-		t.Cert, err = os.ReadFile(t.CertPath)
-		if err != nil {
-			return nil, err
-		}
+	return t.Key, nil
+}
+
+func (t *TLS) updateCachedKeyPair() error {
+	cert, err := t.getCert()
+	if err != nil {
+		return err
+	}
+	key, err := t.getKey()
+	if err != nil {
+		return err
 	}
 	if t.Key == nil || t.Cert == nil {
-		return nil, ErrMissingConfig
+		return ErrMissingConfig
 	}
-	tlsCert, err := tls.X509KeyPair(t.Cert, t.Key)
+	tlsCert, err := tls.X509KeyPair(cert, key)
 	if err != nil {
+		return err
+	}
+	t.cachedCert = &tlsCert
+	return nil
+}
+
+func (t *TLS) Config() (_ *tls.Config, err error) {
+	if !t.Enabled {
+		return nil, nil
+	}
+	if t.mtx == nil {
+		t.mtx = &sync.RWMutex{}
+	}
```

Comment:
This actually creates a race condition on the `t.mtx` field. I recommend not using a pointer to a mutex, so that it gets initialized together with the `TLS`. However, in order to avoid copying the mutex, all references to `TLS` need to become a pointer. This is fine, functions already use a pointer. Only the higher level `Config` struct do not yet, modifying them to a pointer shouldn't break anything.